File: service-beacons-node/src/clients/version1/BeaconsGrpcClientV1.ts
const services = require('../../../../src/protos/beacons_v1_grpc_pb');
const messages = require('../../../../src/protos/beacons_v1_pb');

import { FilterParams } from 'pip-services3-commons-node';
import { PagingParams } from 'pip-services3-commons-node';
import { DataPage } from 'pip-services3-commons-node';
import { GrpcClient } from 'pip-services3-grpc-node';

import { BeaconV1 } from '../../data/version1/BeaconV1';
import { BeaconsGrpcConverterV1 } from '../../services/version1/BeaconsGrpcConverterV1';
import { IBeaconsClientV1 } from './IBeaconsClientV1';

export class BeaconsGrpcClientV1 extends GrpcClient implements IBeaconsClientV1 {
    public constructor() {
        super(services.BeaconsV1Client);
    }

    public getBeacons(correlationId: string, filter: FilterParams, paging: PagingParams,
        callback: (err: any, page: DataPage<BeaconV1>) => void): void {

        let request = new messages.BeaconV1PageRequest();
        BeaconsGrpcConverterV1.setMap(request.getFilterMap(), filter);
        request.setPaging(BeaconsGrpcConverterV1.fromPagingParams(paging));

        let timing = this.instrument(correlationId, 'beacons.get_beacons');

        this.call(
            'get_beacons',
            correlationId,
            request,
            (err, response) => {
                timing.endTiming();

                if (err) {
                    callback(err, null);
                    return;
                }

                let page = BeaconsGrpcConverterV1.toBeaconsPage(response.getPage());
                err = BeaconsGrpcConverterV1.toError(response.getError());
                callback(err, page)
            }
        );
    }

    public getBeaconById(correlationId: string, beaconId: string,
        callback: (err: any, beacon: BeaconV1) => void): void {

        let request = new messages.BeaconV1IdRequest()
        request.setBeaconId(beaconId)
        let timing = this.instrument(correlationId, 'beacons.get_beacon_by_id');
        this.call(
            'get_beacon_by_id',
            correlationId,
            request,
            (err, response) => {
                timing.endTiming();
                if (err) {
                    callback(err, null);
                }
                let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());
                err = BeaconsGrpcConverterV1.toError(response.getError());
                callback(err, beacon);
            }
        );
    }

    public getBeaconByUdi(correlationId: string, udi: string,
        callback: (err: any, beacon: BeaconV1) => void): void {

        let request = new messages.BeaconV1UdiRequest()
        request.setBeaconUdi(udi)

        let timing = this.instrument(correlationId, 'beacons.get_beacon_by_udi');
        this.call(
            'get_beacon_by_udi',
            correlationId,
            request,
            (err, response) => {
                timing.endTiming();
                if (err) {
                    callback(err, null);
                }
                let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());
                err = BeaconsGrpcConverterV1.toError(response.getError());
                callback(err, beacon);
            }
        );
    }

    public calculatePosition(correlationId: string, siteId: string, udis: string[],
        callback: (err: any, position: any) => void): void {
        let request = new messages.BeaconV1PositionRequest()
        request.setSiteId(siteId)
        request.setUdisList(udis)
        let timing = this.instrument(correlationId, 'beacons.calculate_position');
        this.call(
            'calculate_position',
            correlationId,
            request,
            (err, response) => {
                timing.endTiming();
                if (err) {
                    callback(err, null);
                }
                let position = BeaconsGrpcConverterV1.toPosition(response.getPoint());
                err = BeaconsGrpcConverterV1.toError(response.getError());
                callback(err, position);
            }
        );
    }

    public createBeacon(correlationId: string, beacon: BeaconV1,
        callback: (err: any, beacon: BeaconV1) => void): void {

        let request = new messages.BeaconV1ObjectRequest();
        request.setBeacon(BeaconsGrpcConverterV1.fromBeacon(beacon));

        let timing = this.instrument(correlationId, 'beacons.create_beacon');

        this.call(
            'create_beacon',
            correlationId,
            request,
            (err, response) => {
                timing.endTiming();
                if (err) {
                    callback(err, null);
                }
                let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());
                err = BeaconsGrpcConverterV1.toError(response.getError());
                callback(err, beacon);
            }
        );
    }

    public updateBeacon(correlationId: string, beacon: BeaconV1,
        callback: (err: any, beacon: BeaconV1) => void): void {

        let request = new messages.BeaconV1ObjectRequest();
        request.setBeacon(BeaconsGrpcConverterV1.fromBeacon(beacon));

        let timing = this.instrument(correlationId, 'beacons.update_beacon');

        this.call(
            'update_beacon',
            correlationId,
            request,
            (err, response) => {
                timing.endTiming();
                if (err) {
                    callback(err, null);
                }
                let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());
                err = BeaconsGrpcConverterV1.toError(response.getError());
                callback(err, beacon);
            }
        );
    }

    public deleteBeaconById(correlationId: string, beaconId: string,
        callback: (err: any, beacon: BeaconV1) => void): void {

        let request = new messages.BeaconV1IdRequest();
        request.setBeaconId(beaconId);

        let timing = this.instrument(correlationId, 'beacons.delete_beacon_by_id');

        this.call(
            'delete_beacon_by_id',
            correlationId,
            request,
            (err, response) => {
                timing.endTiming()
                if (err) {
                    callback(err, null);
                }
                let beacon = BeaconsGrpcConverterV1.toBeacon(response.getBeacon());
                err = BeaconsGrpcConverterV1.toError(response.getError());
                callback(err, beacon);
            }
        );
    }
}

File: service-beacons-dotnet/src/Client/Clients/Version1/BeaconsGrpcClientV1.cs
using Nov.MaxSamples.Beacons.Data.Version1;
using BeaconsV1;
using PipServices3.Commons.Data;
using PipServices3.Commons.Data.Mapper;
using PipServices3.Commons.Errors;
using PipServices3.Grpc.Clients;
using System;
using System.Threading.Tasks;
using BeaconV1 = Nov.MaxSamples.Beacons.Data.Version1.BeaconV1;
using PagingParams = PipServices3.Commons.Data.PagingParams;

namespace Nov.MaxSamples.Beacons.Clients.Version1
{
	public class BeaconsGrpcClientV1 : GrpcClient, IBeaconsClientV1
	{
		public BeaconsGrpcClientV1()
			: base("beacons_v1")
		{
		}

		public async Task<CenterObjectV1> CalculatePositionAsync(string correlationId, string siteId, string[] udis)
		{
			correlationId ??= IdGenerator.NextLong();

			var methodName = "calculate_position";
			var timing = Instrument(correlationId, methodName);
			try
			{
				var request = new BeaconV1PositionRequest
				{
					CorrelationId = correlationId,
					SiteId = siteId
				};
				request.Udis.AddRange(udis);

				var response = await CallAsync<BeaconV1PositionRequest, BeaconV1PositionReply>(methodName, request);

				if (response != null)
				{
					HandleError(response.Error);
					return BeaconsGrpcConverterV1.ToPosition(response.Point);
				}

				return null;
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				throw ex;
			}
			finally
			{
				timing.EndTiming();
			}
		}

		public async Task<BeaconV1> CreateBeaconAsync(string correlationId, BeaconV1 beacon)
		{
			correlationId ??= IdGenerator.NextLong();

			var methodName = "create_beacon";
			var timing = Instrument(correlationId, methodName);
			try
			{
				var request = new BeaconV1ObjectRequest
				{
					CorrelationId = correlationId,
					Beacon = BeaconsGrpcConverterV1.FromBeacon(beacon)
				};

				var response = await CallAsync<BeaconV1ObjectRequest, BeaconV1ObjectReply>(methodName, request);

				if (response != null)
				{
					HandleError(response.Error);
					return BeaconsGrpcConverterV1.ToBeacon(response.Beacon);
				}

				return null;
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				throw ex;
			}
			finally
			{
				timing.EndTiming();
			}
		}

		public async Task<BeaconV1> UpdateBeaconAsync(string correlationId, BeaconV1 beacon)
		{
			correlationId ??= IdGenerator.NextLong();

			var methodName = "update_beacon";
			var timing = Instrument(correlationId, methodName);
			try
			{
				var request = new BeaconV1ObjectRequest
				{
					CorrelationId = correlationId,
					Beacon = BeaconsGrpcConverterV1.FromBeacon(beacon)
				};

				var response = await CallAsync<BeaconV1ObjectRequest, BeaconV1ObjectReply>(methodName, request);

				if (response != null)
				{
					HandleError(response.Error);
					return BeaconsGrpcConverterV1.ToBeacon(response.Beacon);
				}

				return null;
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				throw ex;
			}
			finally
			{
				timing.EndTiming();
			}
		}

		public async Task<BeaconV1> DeleteBeaconByIdAsync(string correlationId, string id)
		{
			correlationId ??= IdGenerator.NextLong();

			var methodName = "delete_beacon_by_id";
			var timing = Instrument(correlationId, methodName);
			try
			{
				var request = new BeaconV1IdRequest
				{
					CorrelationId = correlationId,
					BeaconId = id
				};

				var response = await CallAsync<BeaconV1IdRequest, BeaconV1ObjectReply>(methodName, request);

				if (response != null)
				{
					HandleError(response.Error);
					return BeaconsGrpcConverterV1.ToBeacon(response.Beacon);
				}

				return null;
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				throw ex;
			}
			finally
			{
				timing.EndTiming();
			}
		}

		public async Task<BeaconV1> GetBeaconByIdAsync(string correlationId, string id)
		{
			correlationId ??= IdGenerator.NextLong();

			var methodName = "get_beacon_by_id";
			var timing = Instrument(correlationId, methodName);
			try
			{
				var request = new BeaconV1IdRequest
				{
					CorrelationId = correlationId,
					BeaconId = id
				};

				var response = await CallAsync<BeaconV1IdRequest, BeaconV1ObjectReply>(methodName, request);

				if (response != null)
				{
					HandleError(response.Error);
					return BeaconsGrpcConverterV1.ToBeacon(response.Beacon);
				}

				return null;
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				throw ex;
			}
			finally
			{
				timing.EndTiming();
			}
		}

		public async Task<BeaconV1> GetBeaconByUdiAsync(string correlationId, string udi)
		{
			correlationId ??= IdGenerator.NextLong();

			var methodName = "get_beacon_by_udi";
			var timing = Instrument(correlationId, methodName);
			try
			{
				var request = new BeaconV1UdiRequest
				{
					CorrelationId = correlationId,
					BeaconUdi = udi
				};

				var response = await CallAsync<BeaconV1UdiRequest, BeaconV1ObjectReply>(methodName, request);

				if (response != null)
				{
					HandleError(response.Error);
					return BeaconsGrpcConverterV1.ToBeacon(response.Beacon);
				}

				return null;
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				throw ex;
			}
			finally
			{
				timing.EndTiming();
			}
		}

		public async Task<DataPage<BeaconV1>> GetBeaconsAsync(string correlationId, FilterParams filter, PagingParams paging)
		{
			correlationId ??= IdGenerator.NextLong();

			var methodName = "get_beacons";
			var timing = Instrument(correlationId, methodName);
			try
			{
				var request = new BeaconV1PageRequest
				{
					CorrelationId = correlationId,
					Paging = BeaconsGrpcConverterV1.FromPaging(paging),
				};

				BeaconsGrpcConverterV1.SetMap(request.Filter, filter);

				var response = await CallAsync<BeaconV1PageRequest, BeaconV1PageReply>(methodName, request);

				if (response != null)
				{
					HandleError(response.Error);
					return BeaconsGrpcConverterV1.ToBeaconsPage(response.Page);
				}

				return null;
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				throw ex;
			}
			finally
			{
				timing.EndTiming();
			}
		}

		private void HandleError(BeaconsV1.ErrorDescription error)
		{
			// Handle error response
			if (error != null)
			{
				var errorDescription = ObjectMapper.MapTo<PipServices3.Commons.Errors.ErrorDescription>(error);
				throw ApplicationExceptionFactory.Create(errorDescription);
			}
		}
	}
}
